tests/js/plugin/enhancedanchoring-test.coffee

```diff
@@ -219,16 +226,52 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
       assert.include hls, anchor2.highlight[anchor2.startPage]
       assert.notInclude hls, anchor3.highlight[anchor3.startPage]
 
+  describe 'Anchor.scrollIntoView()', ->
+    it 'calls scrollIntoView() on the highlight', ->
+      am = createAnchoringManager()
+      ann = createTestAnnotation "a1"
+      anchor = am.createAnchor(ann, ann.target[0]).result
+      anchor.scrollIntoView().then ->
+
+        assert.called anchor.highlight[anchor.startPage].scrollIntoView
+
   describe 'two-phased anchoring', ->
 
+    # Simple lazy rendering document simulation for testing,
+    # which emulates the user movement prediction (and page rendering)
+    # behavior of PDF.js
     class DummyDocumentAccess
 
       @applicable: -> true
 
       isPageMapped: (index) -> index in @_rendered
+      getPageIndex: -> @currentIndex
+
+      setPageIndex: (index) ->
+
+        # Do this when the wanted page is ready
+        onPageReady = =>
+          # Try to find out where will the reader proceed
+          prediction = if @currentIndex < index
+            index + 1  # If we went forward, we'll want the next page, too
+          else
+            index - 1  # if we want backward, we'll want the prev. page, too
+
+          # Set the page index where we want to be
+          @currentIndex = index
+
+          # Render the predicted next page
+          renderPage(this, prediction) unless @isPageMapped prediction
+
+        # Try to move where we need to be
+        if @isPageMapped index
+          onPageReady()
+        else
+          renderPage(this, index).then onPageReady
```

Comment:
There's way too much logic in here that's being tested. For the following tests we should be able to stub `isPageMapped` and `getPageIndex` to return explicit values and then assert `setPageIndex` or `scrollIntoView` is called with correct arguments.